lib/puppet/configurer.rb

```diff
@@ -55,13 +56,20 @@ def initialize(factory = Puppet::Configurer::DownloaderFactory.new)
   # Get the remote catalog, yo.  Returns nil if no catalog can be found.
   def retrieve_catalog(query_options)
     query_options ||= {}
-    # First try it with no cache, then with the cache.
-    unless (Puppet[:use_cached_catalog] and result = retrieve_catalog_from_cache(query_options)) or result = retrieve_new_catalog(query_options)
-      if ! Puppet[:usecacheonfailure]
-        Puppet.warning "Not using cache on failed catalog"
-        return nil
+    if (Puppet[:use_cached_catalog] && result = retrieve_catalog_from_cache(query_options))
+      @cached_catalog_status = 'used because use_cached_catalog was set'
+    else
+      result = retrieve_new_catalog(query_options)      
+
+      if !result 
+        if ! Puppet[:usecacheonfailure]
+          Puppet.warning "Not using cache on failed catalog"
+          return nil
+        end
+
+        @cached_catalog_status = 'used because new catalog failed to compile and usecacheonfailure was set'
+        result = retrieve_catalog_from_cache(query_options)
```

Comment:
Not sure if we really specified this or not, but should we set `cached_catalog_status` here only if we successfully retrieved a catalog from the cache? I.e. saying "used" if we couldn't use it ... feels wrong.